syntax = "proto3";

package webtalk;

import "google/protobuf/timestamp.proto";

service Webtalk {
  rpc Login(LoginRequest) returns (LoginResponse) {}
  rpc Logout(LogoutRequest) returns (LogoutResponse) {}
  rpc UpdateMessage(UpdateMessageRequest) returns (UpdateMessageResponse) {}
  rpc UpdateMessages(stream UpdateMessageRequest) returns (stream UpdateMessageResponse) {}
  rpc ReceiveMessages(ReceiveMessageRequest) returns (stream ReceiveMessageResponse) {}
  rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse) {}
}

message User {
  string nick = 1;
  int64 uid = 2;
}

message SessionToken {
  bytes token = 1;
}

message LoginRequest {
  string nick = 1;
}

message LoginResponse {
  SessionToken session = 1;
  User user = 2;
}

message LogoutRequest {
  SessionToken session = 1;
}

message LogoutResponse {
}

message UpdateMessageRequest {
  SessionToken session = 1;
  Message msg = 2;
}

message UpdateMessageResponse {
  Message msg = 1;
}

message ReceiveMessageRequest {
  SessionToken session = 1;
}

message ReceiveMessageResponse {
  Message msg = 1;
}

message Message {
  int64 msg_id = 1;
  string text = 2;
  User author = 3;
  bool finalized = 4;
  google.protobuf.Timestamp created_time = 5;
  google.protobuf.Timestamp finalized_time = 6;
}

message HeartbeatRequest {
  SessionToken session = 1;
}

message HeartbeatResponse {
}
